doc: add tp-lib stage 1 proposal (#177)
OpenRailAssociation/technical-committee: project-proposals/stage-1/Trainpositioning.md

## What is the project's name?

Trainpositioning lib

## Describe the project. What does the project do, why is it valuable, where does it come from?

The library is used in production by Infrabel. It allows to post process the GNSS location as measured by a train. At Infrabel we use it to improve the positioning of measurement vehicles and thus of the measured values of the track, catenary, signaling, ...

The library works by taking into account:

- the navigability of the topology, you cannot jump from one track to the other
- a train path as registered by the traffic management or the signaling system

## Who are the maintainers of the project (these will be the primary contacts for the OpenRail Association)?

Infrabel:

- glenn.vancalster@infrabel.be
- kevin.goos@infrabel.be
- ilya.quateau@infrabel.be
- mathias.vandenauweele@infrabel.be

## Which organizations are sponsoring/contributing to the project?

Infrabel

## Where is the code hosted?

At the moment, on a gitlab server at Infrabel. We want to opensource it immediately under the openrail organisation.

## Which exact repositories do you intend to transfer to the GitHub organization of the OpenRail Association?

> Originally, the repository contains a JSON API. We will split of the logic from the API and put it in a seperate repository. The API wrapper is not very usefull to opensource, so we will only opensource the library repository.

Reviewer: ```suggestion
Originally, the repository contains a JSON API. We will split of the logic from the API and put it in a separate repository. The API wrapper is not very usefull to opensource, so we will only opensource the library repository.

```

## What is the project's main license?

We propose MIT License or the Apache License 2.0 but want to have the advise of the technical committee

> ## What other licenses does the project use, e.g. for included 3rd party code?

Reviewer: ```suggestion
## What other licenses does the project use, e.g., for included 3rd-party code?

```

> We are still in the progress of simplifying the libary, but at the moment, these packages are used with their respective licenses:

Reviewer: ```suggestion
We are still in the progress of simplifying the library, but at the moment, these packages are used with their respective licenses:

```

- Microsoft.VisualStudio.Azure.Containers.Tools.Targets: [Microsoft specific license](https://www.nuget.org/packages/Microsoft.VisualStudio.Azure.Containers.Tools.Targets/1.21.0/License)
- NetTopologySuite: BSD-3-Clause License
- NodaTime: Apache License 2.0
- Npgsql.NetTopologySuite: PostgreSQL License
- Serilog: Apache License 2.0
- Serilog.Sinks.Console: Apache License 2.0
- DotSpatial.Projections: MIT License

We don't immediately see any major issues.

## Are any trademarks associated with the project?

No

## Does the project have a web site? Where is it? Are you ok with moving it to be hosted by the OpenRail Association?

> No, no website. We will forsee documentation and a seperate .net project to run tests, both will also be open sourced.

Reviewer: ```suggestion
No, no website. We will foresee documentation and a separate .net project to run tests, both will also be open sourced.

```
see:
https://www.merriam-webster.com/dictionary/forsee

- PM: glenn.vancalster@infrabel.be
- Dev: kevin.goos@infrabel.be
- Dev: ilya.quateau@infrabel.be
- Prev PM / Dev: mathias.vandenauweele@infrabel.be

## How is it decided if and when a pull request is merged?

Depends on the pull request, generally the project leader decides.

## How can someone become a committer or a maintainer to/of the project?

After discussion with the project leader and current development team.

## How is development of the project planned and organized? Is this transparent to the public?

Using git issues & epics/stories.

## What is the project's roadmap?

After the project refactoring and extensive testing, no further development is planned.

## What other organizations in the world should be interested in this project?

All organisations that use measurement trains and process the measurements. But the project can also be used as a guideline for organisations that want to post processes location measurements for other domains that include topology and navigability.

## Why would this project be a good candidate for inclusion in the OpenRail Association?

Because all infra managers do this and should better do it together.

> ## Are there competing products or projects? If there are please explain how the proposed projects differentiates.

Reviewer: ```suggestion
## Are there competing products or projects? If there are, please explain how the proposed projects differentiates.

```

Apache License 2.0

## What other licenses does the project use, e.g., for included 3rd-party code?

We are still in the progress of simplifying the library, but at the moment, these packages are used with their respective licenses:

- NetTopologySuite: BSD-3-Clause License
- NodaTime: Apache License 2.0
- Npgsql.NetTopologySuite: PostgreSQL License
- Serilog: Apache License 2.0
- Serilog.Sinks.Console: Apache License 2.0
- DotSpatial.Projections: MIT License

We don't immediately see any major issues.

## Are any trademarks associated with the project?

No

## Does the project have a web site? Where is it? Are you ok with moving it to be hosted by the OpenRail Association?

No, no website. We will foresee documentation and a separate .net project to run tests, both will also be open sourced.

## What are the communication channels the project uses (such as mailing lists, Slack, IRC, etc.)?

Git issues

## What is the project's leadership team and decision-making process?

- PM: glenn.vancalster@infrabel.be
- Dev: kevin.goos@infrabel.be
- Dev: ilya.quateau@infrabel.be
- Prev PM / Dev: mathias.vandenauweele@infrabel.be

## How is it decided if and when a pull request is merged?

> Depends on the pull request, generally the project leader decides.

Author: Yes, PM = Project leader. 

>  Is the PM a technical role?

That depends on your definition of 'technical'

> Depends on the pull request, generally the project leader decides.

Reviewer 2: In this context I would define "technical" as "will work with the project on GitHub and is at least able to do commits to the git repository".

> Depends on the pull request, generally the project leader decides.

Author: That's going to be the case yes. I assume that the actual merging (pushing the button) will be performed by a developer.

## How can someone become a committer or a maintainer to/of the project?

After discussion with the project leader and current development team.

## How is development of the project planned and organized? Is this transparent to the public?

Using git issues & epics/stories.

## What is the project's roadmap?

> After the project refactoring and extensive testing, no further development is planned.

Author: > What does that mean in terms of resources allocated to the project?

The TP-lib is part of a bigger project that has a dedicated team of 4 and 5 engineers. There is always support for the project for internal needs. 

> Are there areas where other users will likely need further development?

That's difficult for me to answer. I don't know the needs of "other users".

## What other organizations in the world should be interested in this project?

> All organisations that use measurement trains and process the measurements. But the project can also be used as a guideline for organisations that want to post processes location measurements for other domains that include topology and navigability.

Author: Measurements done by a measurement train need to be linked to a track position. The TP-lib helps with this aspect of positioning. 

A measurement train will generally measure a value and save that value either in a kind of ledger together with other measurements such as gnss position. Or it could save the value together with the timestamp in a seperate file. _In a post processing step, the gnss position is improved and positioned on the track topology_. Afterwards, the measured value is positioned on the track topology.
The TP-lib helps with the italic step.

## Why would this project be a good candidate for inclusion in the OpenRail Association?

Because all infra managers do this and should better do it together.

## Are there competing products or projects? If there are, please explain how the proposed projects differentiates.

Not that we know of.

## What standards does the project implement or rely on? How are they related to other existing standards?

> Topology and navigability is also defined by several other railway related standards. But they are very loosely related.

Author: I don't understand this question. It's a dotnet library so the "form" is a dotnet object of a dotnet class inside a dotnet list object.

> Topology and navigability is also defined by several other railway related standards. But they are very loosely related.

Author: Yes, please review the presentation given on the BoD

![image](https://github.com/user-attachments/assets/43cb7da7-ec29-47ee-8f96-c1c8c60de1a2)

![image](https://github.com/user-attachments/assets/22a218ea-8517-44c1-85e3-09827f467078)

> Topology and navigability is also defined by several other railway related standards. But they are very loosely related.

Reviewer 2: So there is no API for reading data about topology and train paths from a standard format?

> Topology and navigability is also defined by several other railway related standards. But they are very loosely related.

Reviewer 2: Not all members of the Technical Committee have access to the BoD materials. So if there is any more info about the context and how to understand the diagrams please add them here as well.

> Topology and navigability is also defined by several other railway related standards. But they are very loosely related.

Author: To my knowledge, the only standard format for topology that exists at the moment is railML. Besides the moral objections from some of the Openrail members, the TP-lib currently does not support working with this format.

RCM-DX also allows to define the topology but this definition is not sufficiently railway independent (still heavily SBB). We are working on making that more open and according to RSM/RTM standards.

Hopefully we will be able to implement a connector with the RINF by the start of 2026, when all IM's need to publish their microtopology according to EU legislation. 

At the moment, al data transfer into the library is only defined as an "Interface". That means that the necessary functions and their signature (properties input and output) are defined but the actual work isn't. This allows every user of the library to implement their own calls. Be it to a knowledge graph, a database, a file or something else.

> Topology and navigability is also defined by several other railway related standards. But they are very loosely related.

Author: RCM-DX is a data format. The RCM-DX file can contain GNSS data. In case topological locations are also requested, TP-Lib could be used to read data out of the file and write its result back into the RCM-DX file. TP-lib is currently not able to read/write to RCM-DX but an implementation could be developed.
